src/httpserver.cpp

```diff
@@ -40,6 +41,55 @@
 /** Maximum size of http request (request line + headers) */
 static const size_t MAX_HEADERS_SIZE = 8192;
 
+class ConnectionLimiter
+{
+public:
+    ConnectionLimiter(std::vector<evconnlistener*> listeners, unsigned int limit) : m_limit(limit), m_listeners(std::move(listeners))
+    {
+        assert(m_limit > 0);
+    }
+    void AddConnection(evutil_socket_t fd)
+    {
+        // Disable socket accepting if adding this connection puts us equal to the limit
+        if (!Interrupted() && m_sockets.insert(fd).second && m_sockets.size() == m_limit) {
+            LogPrint(BCLog::HTTP, "Suspending new connections");
+            for (const auto& listener : m_listeners) {
+                evconnlistener_disable(listener);
+            }
+        }
+    }
+    void RemoveConnection(evutil_socket_t fd)
+    {
+        // Re-enable socket accepting if removing this connection brings us
+        // back down under the limit
+        if (m_sockets.erase(fd) && m_sockets.size() + 1 == m_limit && !Interrupted()) {
+            LogPrint(BCLog::HTTP, "Resuming new connections\n");
+            for (const auto& listener : m_listeners) {
+                evconnlistener_enable(listener);
+            }
+        }
+    }
+    bool IsReady() const
+    {
+        return m_sockets.size() < m_limit && !Interrupted();
+    }
+    void Interrupt()
+    {
+        m_interrupted.store(true, std::memory_order_release);
+    }
+private:
+
+    inline bool Interrupted() const
+    {
+        return m_interrupted.load(std::memory_order_acquire);
+    }
+
+    const unsigned int m_limit;
+    std::vector<evconnlistener*> m_listeners;
+    std::set<evutil_socket_t> m_sockets;
```

Comment:
Not that it probably matters in most configurations where the fd limit is ~1K, but making this an unordered_set would be slightly beneficial in the high fd limit cases.

Reply:
Sure